src/nvim/drawline.c

```diff
@@ -960,7 +960,7 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
   filler_todo = filler_lines;
 
   // Cursor line highlighting for 'cursorline' in the current window.
-  if (lnum == wp->w_cursor.lnum) {
+  if (lnum == wp->w_cursorline) {
     // Do not show the cursor line in the text when Visual mode is active,
     // because it's not clear what is selected then.
     if (wp->w_p_cul && !(wp == curwin && VIsual_active)
```

Comment:
Now checking for `w_p_cul` first makes more sense here (and is the same as Vim) because `w_cursorline` is not set if `w_p_cul` is false.
```suggestion
  if (wp->w_p_cul && lnum == wp->w_cursorline) {
    // Do not show the cursor line in the text when Visual mode is active,
    // because it's not clear what is selected then.
    if (!(wp == curwin && VIsual_active)
```

Reply:
Should it just be a single if block, i.e. reduce the nesting level?

Reply:
I guess yet.

Reply:
```diff
+++ b/src/nvim/drawline.c
@@ -960,20 +960,17 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
   filler_todo = filler_lines;

   // Cursor line highlighting for 'cursorline' in the current window.
-  if (lnum == wp->w_cursorline) {
-    // Do not show the cursor line in the text when Visual mode is active,
-    // because it's not clear what is selected then.
-    if (wp->w_p_cul && !(wp == curwin && VIsual_active)
-        && wp->w_p_culopt_flags != CULOPT_NBR) {
-      cul_screenline = (wp->w_p_wrap
-                        && (wp->w_p_culopt_flags & CULOPT_SCRLINE));
-      if (!cul_screenline) {
-        apply_cursorline_highlight(wp, lnum, &line_attr, &cul_attr, &line_attr_lowprio);
-      } else {
-        margin_columns_win(wp, &left_curline_col, &right_curline_col);
-      }
-      area_highlighting = true;
+  // Do not show the cursor line in the text when Visual mode is active,
+  // because it's not clear what is selected then.
+  if (wp->w_p_cul && wp->w_p_culopt_flags != CULOPT_NBR
+      && lnum == wp->w_cursorline && !(wp == curwin && VIsual_active) ) {
+    cul_screenline = (wp->w_p_wrap && (wp->w_p_culopt_flags & CULOPT_SCRLINE));
+    if (!cul_screenline) {
+      apply_cursorline_highlight(wp, lnum, &line_attr, &cul_attr, &line_attr_lowprio);
+    } else {
+      margin_columns_win(wp, &left_curline_col, &right_curline_col);
     }
+    area_highlighting = true;
   }

   SignTextAttrs sattrs[SIGN_SHOW_MAX];  // sign attributes for the sign column
```

Reply:
I guess yet. So
```diff
diff --git a/src/nvim/drawline.c b/src/nvim/drawline.c
index a4b9e552a..25db3c7d1 100644
--- a/src/nvim/drawline.c
+++ b/src/nvim/drawline.c
@@ -960,20 +960,18 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
   filler_todo = filler_lines;
 
   // Cursor line highlighting for 'cursorline' in the current window.
-  if (lnum == wp->w_cursorline) {
-    // Do not show the cursor line in the text when Visual mode is active,
-    // because it's not clear what is selected then.
-    if (wp->w_p_cul && !(wp == curwin && VIsual_active)
-        && wp->w_p_culopt_flags != CULOPT_NBR) {
-      cul_screenline = (wp->w_p_wrap
-                        && (wp->w_p_culopt_flags & CULOPT_SCRLINE));
-      if (!cul_screenline) {
-        apply_cursorline_highlight(wp, lnum, &line_attr, &cul_attr, &line_attr_lowprio);
-      } else {
-        margin_columns_win(wp, &left_curline_col, &right_curline_col);
-      }
-      area_highlighting = true;
+  if (wp->w_p_cul && lnum == wp->w_cursor.lnum
+      // Do not show the cursor line in the text when Visual mode is active,
+      // because it's not clear what is selected then.
+      && !(wp == curwin && VIsual_active)
+      && wp->w_p_culopt_flags != CULOPT_NBR) {
+    cul_screenline = (wp->w_p_wrap && (wp->w_p_culopt_flags & CULOPT_SCRLINE));
+    if (!cul_screenline) {
+      apply_cursorline_highlight(wp, lnum, &line_attr, &cul_attr, &line_attr_lowprio);
+    } else {
+      margin_columns_win(wp, &left_curline_col, &right_curline_col);
     }
+    area_highlighting = true;
   }
 
   SignTextAttrs sattrs[SIGN_SHOW_MAX];  // sign attributes for the sign column
```

Reply:
I guess yes, so
```diff
diff --git a/src/nvim/drawline.c b/src/nvim/drawline.c
index a4b9e552a..b7ebd157b 100644
--- a/src/nvim/drawline.c
+++ b/src/nvim/drawline.c
@@ -960,20 +960,17 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
   filler_todo = filler_lines;
 
   // Cursor line highlighting for 'cursorline' in the current window.
-  if (lnum == wp->w_cursorline) {
-    // Do not show the cursor line in the text when Visual mode is active,
-    // because it's not clear what is selected then.
-    if (wp->w_p_cul && !(wp == curwin && VIsual_active)
-        && wp->w_p_culopt_flags != CULOPT_NBR) {
-      cul_screenline = (wp->w_p_wrap
-                        && (wp->w_p_culopt_flags & CULOPT_SCRLINE));
-      if (!cul_screenline) {
-        apply_cursorline_highlight(wp, lnum, &line_attr, &cul_attr, &line_attr_lowprio);
-      } else {
-        margin_columns_win(wp, &left_curline_col, &right_curline_col);
-      }
-      area_highlighting = true;
+  if (wp->w_p_cul && wp->w_p_culopt_flags != CULOPT_NBR && lnum == wp->w_cursorline
+      // Do not show the cursor line in the text when Visual mode is active,
+      // because it's not clear what is selected then.
+      && !(wp == curwin && VIsual_active)) {
+    cul_screenline = (wp->w_p_wrap && (wp->w_p_culopt_flags & CULOPT_SCRLINE));
+    if (!cul_screenline) {
+      apply_cursorline_highlight(wp, lnum, &line_attr, &cul_attr, &line_attr_lowprio);
+    } else {
+      margin_columns_win(wp, &left_curline_col, &right_curline_col);
     }
+    area_highlighting = true;
   }
 
   SignTextAttrs sattrs[SIGN_SHOW_MAX];  // sign attributes for the sign column
```